### ISSUE TITLE
GLSL 4.10 is not supported

### ISSUE THREAD

**Issue by SteveZhao13** [dilevin/computer-graphics-shader-pipeline#9]
![issue](https://user-images.githubusercontent.com/20450684/54057126-04b4e100-41c0-11e9-8b55-e75c01249d3c.jpg)
As said in the title and showed in the picture, I am trying to run the program without changing anything (./shaderpipeline ../data/test-01.json). It should be a icosahedron but something went wrong.
It seems that this issue is because my hardware? I wonder if there is some ways for me to solve it or do I need to use another laptop?

**Comment by ACalza** [comment]
Ubuntu doesn't install openGL 4.1 by default according to this:

https://askubuntu.com/questions/713586/steam-and-opengl-4-1 

tldr; 

```bash
sudo add-apt-repository ppa:oibaf/graphics-drivers
sudo apt-get update
sudo apt-get upgrade`

**Comment by SteveZhao13** [comment]
> Ubuntu doesn't install openGL 4.1 by default according to this:
> 
> https://askubuntu.com/questions/713586/steam-and-opengl-4-1
> 
> tldr;
> 
> ```shell
> sudo add-apt-repository ppa:oibaf/graphics-drivers
> sudo apt-get update
> sudo apt-get upgrade`
> ```

Hi my friend, I have tried your solution, but still the same error. Do I need to install other package? Thansk a lot.

**Comment by ACalza** [comment]
Screen shot of error?

**Comment by SteveZhao13** [comment]
> Screen shot of error?

![issue2](https://user-images.githubusercontent.com/20450684/54059619-c9b6ab80-41c7-11e9-9402-c6b38e3718d3.png)
the same as before

**Comment by ACalza** [comment]
I meant the screenshot of the installation from: 

```shell
sudo add-apt-repository ppa:oibaf/graphics-drivers
sudo apt-get update
sudo apt-get upgrade

**Comment by rin-23** [comment]
There should be plenty of resources online on how to install opengl4.0 on ubuntu.

**Comment by SteveZhao13** [comment]
> I meant the screenshot of the installation from:
> 
> ```shell
> sudo add-apt-repository ppa:oibaf/graphics-drivers
> sudo apt-get update
> sudo apt-get upgrade
> ```

![issue3](https://user-images.githubusercontent.com/20450684/54064600-c4675a00-41e3-11e9-8ca2-0314fe2aa7c8.png)
![issue4](https://user-images.githubusercontent.com/20450684/54064601-c7624a80-41e3-11e9-86dd-ad47312fb8cf.png)
Something like that. Cause I have restarted my VM so it shows that no upgraded

**Comment by WChrisK** [comment]
What is your top level OS? What OpenGL version does it support?

There's likely a few things standing in your way. First is that if you're using something like a mac, or a laptop with an archaic version of OpenGL (which happened to be anything with an integrated chip in around the 2010 years, specifically Intel chips back then sucked and the most you'd get from them is OpenGL 3.3), that makes it a lot harder. If it is the case that the hardware cannot support it, then somehow getting later verisons of OpenGL won't help. VirtualBox may also be bound by whatever the hosts capabilities are as well. There's plenty of things that could cause you grief here.

The best bet is to do this outside of VirtualBox.

If it still doesn't work on your main computer, you may want to consider installing Mesa3D's latest version since it might possibly give you software emulation of OpenGL... however I have never done this and possibly have no idea what I'm talking about (idk if it actually does software emulation or depends on your graphics drivers). Then again [mesa's software renderer might be only 3.3](https://en.wikipedia.org/wiki/Mesa_(computer_graphics)#Software_renderer) so I could be very wrong here.

The best thing you could do right now is to tell us what OS you are running VB under, and tell us your OpenGL capabilities of the hardware that your main OS has access to.

**Comment by SteveZhao13** [comment]
> What is your top level OS? What OpenGL version does it support?
> 
> There's likely a few things standing in your way. First is that if you're using something like a mac, or a laptop with an archaic version of OpenGL (which happened to be anything with an integrated chip in around the 2010 years, specifically Intel chips back then sucked and the most you'd get from them is OpenGL 3.3), that makes it a lot harder. If it is the case that the hardware cannot support it, then somehow getting later verisons of OpenGL won't help. VirtualBox may also be bound by whatever the hosts capabilities are as well. There's plenty of things that could cause you grief here.
> 
> The best bet is to do this outside of VirtualBox.
> 
> If it still doesn't work on your main computer, you may want to consider installing Mesa3D's latest version since it might possibly give you software emulation of OpenGL... however I have never done this and possibly have no idea what I'm talking about (idk if it actually does software emulation or depends on your graphics drivers). Then again [mesa's software renderer might be only 3.3](https://en.wikipedia.org/wiki/Mesa_(computer_graphics)#Software_renderer) so I could be very wrong here.
> 
> The best thing you could do right now is to tell us what OS you are running VB under, and tell us your OpenGL capabilities of the hardware that your main OS has access to.

![issue](https://user-images.githubusercontent.com/20450684/54135352-04029180-43f0-11e9-9d8f-5e2aba76b934.png)
my top level OS is win10 and as the pic shows above, it support openGL 4.3, so...

**Comment by WChrisK** [comment]
You should try getting the latest version of Mesa3D on your VM and see if that can work. Your hardware capabilities are fine, but it may be the case that you'll be stuck doing development on Windows for this assignment if your VM wants to troll you. Your best bet is to try to get the latest Mesa3D version and hope VB lets the implementation use the hosts GPU capabilities. Sadly there's no guarantee with this stuff and I've had headaches trying to get issues like these to resolve in the past.

Otherwise you'll be in Bahen or working on Windows ;)
